File: src/go/PhantasyIslandGoRemoteControl/radio/wait_token.go
package radio

import (
	"context"
	"sync"
	"time"
)

// WaitToken 一次请求对应的小状态机。
type WaitToken struct {
	WaitCmd           string
	TimeBaseID        int64
	Response          map[string]interface{}
	ProcessedResponse interface{}
	doneChan          chan struct{}
	mu                sync.Mutex
	postProcessor     func(map[string]interface{}) interface{}
}

func NewWaitToken(waitCmd string, timeBaseID int64) *WaitToken {
	return &WaitToken{
		WaitCmd:    waitCmd,
		TimeBaseID: timeBaseID,
		doneChan:   make(chan struct{}),
	}
}

// Done 令牌是否已收到响应。
func (wt *WaitToken) Done() bool {
	select {
	case <-wt.doneChan:
		return true
	default:
		return false
	}
}

// SetPostProcessor 注册后处理回调。
func (wt *WaitToken) SetPostProcessor(processor func(map[string]interface{}) interface{}) *WaitToken {
	wt.mu.Lock()
	defer wt.mu.Unlock()
	wt.postProcessor = processor
	return wt
}

func (wt *WaitToken) applyPostProcessor(data map[string]interface{}) interface{} {
	wt.mu.Lock()
	defer wt.mu.Unlock()
	if wt.postProcessor != nil {
		wt.ProcessedResponse = wt.postProcessor(data)
	} else {
		wt.ProcessedResponse = data
	}
	return wt.ProcessedResponse
}

// Complete 填充响应并唤醒等待者。
func (wt *WaitToken) Complete(data map[string]interface{}) {
	wt.mu.Lock()
	if wt.Done() {
		wt.mu.Unlock()
		return
	}
	wt.Response = data
	wt.mu.Unlock()

	wt.applyPostProcessor(data)

	close(wt.doneChan)
}

// Wait 阻塞当前线程直到收到响应或超时。
func (wt *WaitToken) Wait(timeout time.Duration) interface{} {
	select {
	case <-wt.doneChan:
		wt.mu.Lock()
		defer wt.mu.Unlock()
		return wt.ProcessedResponse
	case <-time.After(timeout):
		return nil
	}
}

// WaitAsync 异步等待（支持 context）。
func (wt *WaitToken) WaitAsync(ctx context.Context) (interface{}, error) {
	select {
	case <-wt.doneChan:
		wt.mu.Lock()
		defer wt.mu.Unlock()
		return wt.ProcessedResponse, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

File: src/rs/src/radio/wait_token.rs
use std::fmt;
use std::sync::{Arc, Mutex, Condvar};
use std::time::Duration;
use serde_json::Value;

pub struct WaitTokenInner {
    pub wait_cmd: String,
    pub time_base_id: i64,
    pub response: Option<Value>,
    pub done: bool,
}

pub struct WaitToken {
    pub inner: Arc<(Mutex<WaitTokenInner>, Condvar)>,
    pub post_processor: Option<Box<dyn Fn(Value) -> Value + Send + Sync>>,
}

impl fmt::Debug for WaitToken {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let inner = self.inner.0.lock().unwrap();
        f.debug_struct("WaitToken")
            .field("wait_cmd", &inner.wait_cmd)
            .field("time_base_id", &inner.time_base_id)
            .field("done", &inner.done)
            .finish()
    }
}

impl WaitToken {
    pub fn new(wait_cmd: &str, time_base_id: i64) -> Self {
        let inner = WaitTokenInner {
            wait_cmd: wait_cmd.to_string(),
            time_base_id,
            response: None,
            done: false,
        };
        Self {
            inner: Arc::new((Mutex::new(inner), Condvar::new())),
            post_processor: None,
        }
    }

    pub fn set_post_processor<F>(&mut self, processor: F)
    where
        F: Fn(Value) -> Value + Send + Sync + 'static,
    {
        self.post_processor = Some(Box::new(processor));
    }

    pub fn complete(&self, data: Value) {
        let (lock, cvar) = &*self.inner;
        let mut inner = lock.lock().unwrap();
        
        let processed_data = if let Some(ref processor) = self.post_processor {
            processor(data.clone())
        } else {
            data.clone()
        };

        inner.response = Some(processed_data);
        inner.done = true;
        cvar.notify_all();
    }

    pub fn wait(&self, timeout: Duration) -> Option<Value> {
        let (lock, cvar) = &*self.inner;
        let mut inner = lock.lock().unwrap();
        
        let result = cvar.wait_timeout_while(inner, timeout, |inner| !inner.done).unwrap();
        inner = result.0;
        
        inner.response.clone()
    }

    pub fn is_done(&self) -> bool {
        self.inner.0.lock().unwrap().done
    }
}
